### ISSUE TITLE
Monero fork support

### ISSUE THREAD

**Issue by bs3vcenk** [nioroso-x3/xmr-stak-power#9]
Curious about when the algorithm changes will be implemented. Every other major software has it implemented, but the last commit was made 4 months ago here.

**Comment by nioroso-x3** [comment]
I haven't had much time for this project, I'll look into it.

**Comment by nioroso-x3** [comment]
I guess I'll just patch the new xmr-stak, might take a week or two.

**Comment by Balzhur** [comment]
@nioroso-x3, just to be clear, by "new xmr-stak" you mean 2.3.x by fierce-uk?
Cool then, it has some nice improvements.

**Comment by madscientist159** [comment]
@nioroso-x3 Count me in as eager for an update.  V7 is just around the corner and would hate for my P8 machine to go offline....

**Comment by nioroso-x3** [comment]
Ok, it was simpler than I thought.
Hash rate dropped for the normal monero algorithm, I wonder if it is because the functions are larger?
The new port is here: https://github.com/nioroso-x3/xmr-stak

**Comment by Balzhur** [comment]
@nioroso-x3, hashrate dropped from 2130ish to 1800ish for generated config on Power8 machine. But once I changed "be_mode" to "false" it's almost back to 2100ish.

This is an LPAR with 22 Power8 cores, SMT=2, each miner instance running 11 cores/22 threads in parallel:

**xmr-stak-power 1.5.0**
`"low_power_mode" : true, "little_endian_mode" : false`
```
Totals:   2130.1 2128.6 2128.7 H/s
Highest:  2137.3 H/s
```

**xmr-stak 2.3.0**
`"low_power_mode" : 2, "be_mode" : false`
```
Totals (ALL):   2098.6 2098.3 2098.5 H/s
Highest:  2101.1 H/s
```

**Comment by Balzhur** [comment]
oh, and THANKS!

**Comment by gcy000** [comment]
hi guys, it seem that after successful compile and make install without any error from the source code at the new link for the coming monero fork at https://github.com/nioroso-x3/xmr-stak, 
I've got issue running the xmr-stak with error below:
./xmr-stak: error while loading shared libraries: libhwloc.so.5: cannot open shared object file 
Is there anything i might have miss out? As i have no issue running the previous xmr-stak-power for my power8...thks

**Comment by Balzhur** [comment]
@gcy000, configure as following:
`cmake .. -DHWLOC_ENABLE=OFF`

**Comment by gcy000** [comment]
try it seem while performing make error below:
In file included from /xmr-stak/xmrstak/backend/cpu/minethd.cpp:43:0:
/xmr-stak/xmrstak/backend/cpu/autoAdjust.hpp: In member function âbool xmrstak::                                                                                                             cpu::autoAdjust::printConfig()â:
/xmr-stak/xmrstak/backend/cpu/autoAdjust.hpp:41:23: error: âclass jconfâ has no                                                                                                              member named âIsCurrencyMoneroâ
   if(::jconf::inst()->IsCurrencyMonero())
                       ^~~~~~~~~~~~~~~~
/xmr-stak/xmrstak/backend/cpu/autoAdjust.hpp:43:20: error: âMONERO_MEMORYâ was n                                                                                                             ot declared in this scope
    hashMemSizeKB = MONERO_MEMORY / 1024u;
                    ^~~~~~~~~~~~~
/xmr-stak/xmrstak/backend/cpu/autoAdjust.hpp:48:20: error: âAEON_MEMORYâ was not                                                                                                              declared in this scope
    hashMemSizeKB = AEON_MEMORY / 1024u;
                    ^~~~~~~~~~~
CMakeFiles/xmr-stak-backend.dir/build.make:110: recipe for target 'CMakeFiles/xm                                                                                                             r-stak-backend.dir/xmrstak/backend/cpu/minethd.cpp.o' failed
make[2]: *** [CMakeFiles/xmr-stak-backend.dir/xmrstak/backend/cpu/minethd.cpp.o]                                                                                                              Error 1
CMakeFiles/Makefile2:104: recipe for target 'CMakeFiles/xmr-stak-backend.dir/all                                                                                                             ' failed
make[1]: *** [CMakeFiles/xmr-stak-backend.dir/all] Error 2
Makefile:116: recipe for target 'all' failed
make: *** [all] Error 2

Just check do i need to wipe out and re-install from scrap and try it again?

**Comment by Balzhur** [comment]
@gcy000, I guess you should read "how to make" documentation, create proper directory and run configure and compile from there.

**Comment by ethancedrik** [comment]
Yeah I got the same error as @gcy000 compiling on Ubuntu 17.10 following the "Ubuntu" instructions here https://github.com/nioroso-x3/xmr-stak/blob/master/doc/compile_Linux.md
With compilers
ii  g++                                        4:7.2.0-1ubuntu1                             ppc64el      GNU C++ compiler  
ii  g++-7                                      7.2.0-8ubuntu3.2                             ppc64el      GNU C++ compiler  
ii  gcc                                        4:7.2.0-1ubuntu1                             ppc64el      GNU C compiler  
ii  gcc-7                                      7.2.0-8ubuntu3.2                             ppc64el      GNU C compiler  

I don't have IBM AT11 installed, so I wonder if I should try and use that compiler instead of gcc/g++7?

[![https://gyazo.com/bb5987a32526007843b1630bc2946fdf](https://i.gyazo.com/bb5987a32526007843b1630bc2946fdf.png)](https://gyazo.com/bb5987a32526007843b1630bc2946fdf)


I'll try and see if I can figure out what I'm doing wrong, I'm not too experienced with compiling but I was able to do it with xmr-stak-power with no issues so I don't know...

**Comment by nioroso-x3** [comment]
I just tested with gcc 7.2 on centos, it also fails. I fixed the lines and now it compiles, please test.

**Comment by tayore** [comment]
hello, compiled on ubuntu 16 LTS   / gcc version 6.3.1 20170515 (Advance-Toolchain-at10.0) 
But did not work

root@xxx:~/xmr-stak/bin# ./xmr-stak 
./xmr-stak: /opt/at10.0/lib64/power8/libcrypto.so.1.0.0: no version information available (required by ./xmr-stak)
./xmr-stak: /usr/lib/powerpc64le-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.22' not found (required by ./xmr-stak)
./xmr-stak: /opt/at10.0/lib64/power8/libssl.so.1.0.0: no version information available (required by ./xmr-stak)

**Comment by night2417** [comment]
hello make install error
[root@localhost xmr-stak]# make install
Scanning dependencies of target xmr-stak-c
[  3%] Building C object CMakeFiles/xmr-stak-c.dir/xmrstak/backend/cpu/crypto/c_blake256.c.o
[  7%] Building C object CMakeFiles/xmr-stak-c.dir/xmrstak/backend/cpu/crypto/c_groestl.c.o
[ 11%] Building C object CMakeFiles/xmr-stak-c.dir/xmrstak/backend/cpu/crypto/c_jh.c.o
[ 15%] Building C object CMakeFiles/xmr-stak-c.dir/xmrstak/backend/cpu/crypto/c_keccak.c.o
[ 19%] Building C object CMakeFiles/xmr-stak-c.dir/xmrstak/backend/cpu/crypto/c_skein.c.o
[ 23%] Linking C static library bin/libxmr-stak-c.a
[ 23%] Built target xmr-stak-c
Scanning dependencies of target xmr-stak-backend
[ 26%] Building CXX object CMakeFiles/xmr-stak-backend.dir/xmrstak/jconf.cpp.o
[ 30%] Building CXX object CMakeFiles/xmr-stak-backend.dir/xmrstak/version.cpp.o
[ 34%] Building CXX object CMakeFiles/xmr-stak-backend.dir/xmrstak/backend/cpu/hwlocMemory.cpp.o
[ 38%] Building CXX object CMakeFiles/xmr-stak-backend.dir/xmrstak/backend/cpu/jconf.cpp.o
[ 42%] Building CXX object CMakeFiles/xmr-stak-backend.dir/xmrstak/backend/cpu/minethd.cpp.o
In file included from /usr/local/xmr-stak/xmrstak/backend/cpu/minethd.cpp:43:0:
/usr/local/xmr-stak/xmrstak/backend/cpu/autoAdjust.hpp: In member function ‘bool xmrstak::cpu::autoAdjust::printConfig()’:
/usr/local/xmr-stak/xmrstak/backend/cpu/autoAdjust.hpp:41:23: error: ‘class jconf’ has no member named ‘IsCurrencyMonero’
   if(::jconf::inst()->IsCurrencyMonero())
                       ^~~~~~~~~~~~~~~~
/usr/local/xmr-stak/xmrstak/backend/cpu/autoAdjust.hpp:43:20: error: ‘MONERO_MEMORY’ was not declared in this scope
    hashMemSizeKB = MONERO_MEMORY / 1024u;
                    ^~~~~~~~~~~~~
/usr/local/xmr-stak/xmrstak/backend/cpu/autoAdjust.hpp:48:20: error: ‘AEON_MEMORY’ was not declared in this scope
    hashMemSizeKB = AEON_MEMORY / 1024u;
                    ^~~~~~~~~~~
make[2]: *** [CMakeFiles/xmr-stak-backend.dir/xmrstak/backend/cpu/minethd.cpp.o] error 1
make[1]: *** [CMakeFiles/xmr-stak-backend.dir/all] error 2
make: *** [all] error 2


help me

**Comment by ethancedrik** [comment]
@night2417 what you're getting looks exactly like what I was getting

I've removed all compilers from my system (`apt-get remove g++` | `apt-get remove gcc`) and installed IBM AT11 which you can do by 
`wget ftp://ftp.unicamp.br/pub/linuxpatch/toolchain/at/ubuntu/dists/trusty/6976a827.gpg.key`  
`sudo apt-key add 6976a827.gpg.key `  
`nano /etc/apt/sources.list`    
(scroll down to the bottom and add)  
`deb ftp://ftp.unicamp.br/pub/linuxpatch/toolchain/at/ubuntu xenial at11.0`    
(then save and quit)  
`apt-get update`  
`apt-get install advance-toolchain-at11.0-runtime advance-toolchain-at11.0-devel  advance-toolchain-at11.0-perf   advance-toolchain-at11.0-mcore-libs`    

and now cmake is unable to identify the compiler (`No CMAKE_C_COMPILER could be found.` | `No CMAKE_CXX_COMPILER could be found.`) apparently `export PATH=/opt/at11.0/bin:$PATH` doesn't do it so I'll have to try and set that up in the environment path then I'll try compiling again.
`export CXX=/opt/at11.0/bin/g++`  
`export CC=/opt/at11.0/bin/gcc`

**Comment by ethancedrik** [comment]
Yeah after I did the above (compiling with IBM AT11) I got  
[![https://gyazo.com/2f9d359ef505f8740a65c487f4c9cde3](https://i.gyazo.com/2f9d359ef505f8740a65c487f4c9cde3.png)](https://gyazo.com/2f9d359ef505f8740a65c487f4c9cde3)

Entire log is here
[cmakelog.txt](https://github.com/nioroso-x3/xmr-stak-power/files/1885147/cmakelog.txt)

**Comment by Balzhur** [comment]
AT 10.0
```
git clone https://github.com/nioroso-x3/xmr-stak
cd xmr-stak/
mkdir build
cd build/
export LD_LIBRARY_PATH=/usr/lib/powerpc64le-linux-gnu/
export PATH=/opt/at10.0/bin:$PATH
cmake .. -DOpenSSL_ENABLE=OFF -DMICROHTTPD_ENABLE=OFF -DCMAKE_LINK_STATIC=ON
make
```

**Comment by ethancedrik** [comment]
I removed AT11 and installed AT10
then tried 

> `git clone https://github.com/nioroso-x3/xmr-stak`  
> `cd xmr-stak/`  
> `mkdir build`  
> `cd build/`
> `export LD_LIBRARY_PATH=/usr/lib/powerpc64le-linux-gnu/`  
> `export PATH=/opt/at10.0/bin:$PATH`  
> `cmake .. -DOpenSSL_ENABLE=OFF -DMICROHTTPD_ENABLE=OFF -DCMAKE_LINK_STATIC=ON`
> `make`  

and cmake and make worked

then I did `make install` and it looks like it's worked!

**Comment by Balzhur** [comment]
@ethancedrik, I think it worked not because of AT10, but because of `-DOpenSSL_ENABLE=OFF`
:)

**Comment by ethancedrik** [comment]
>@ethancedrik, I think it worked not because of AT10, but because of -DOpenSSL_ENABLE=OFF
:)
That's very likely given my last errors were related to that :)

[![https://gyazo.com/e6271c8238cca8aefb926b64f7540aee](https://i.gyazo.com/e6271c8238cca8aefb926b64f7540aee.png)](https://gyazo.com/e6271c8238cca8aefb926b64f7540aee)

Not sure why I'm getting this when I try and start it up

Edit: cpu.txt needed to be edited and configured, the setup utility didn't do it automatically

Now I've encountered another issue  
`[2018-04-06 18:48:44] : Fast-connecting to pool.supportxmr.com:7777 pool ...`     
`[2018-04-06 18:48:44] : SOCKET ERROR - [pool.supportxmr.com:7777] CONNECT error: GetAddrInfo: Temporary failure in name resolution`    

I can ping pool.supportxmr.com just fine and resolve it, and I've never had this issue before, nothing on the network has changed, I've also tried using port 5555 which I know also worked on xmr-stak-power and still the same issue

**Comment by nioroso-x3** [comment]
I wonder what causes those errors in ubuntu. We dont have any power8 machines with ubuntu here, only Centos7, so I can't really test it.

**Comment by ethancedrik** [comment]
I have installed CentOS 7 on the system and tried compiling and installing, and ran it and it just sits at `[2018-04-06 18:21:44] : Fast-connecting to pool.supportxmr.com:7777 pool ...`  then times out

I wonder if disabling openssl has something to do with it

**Comment by madscientist159** [comment]
Just wanted to weigh in and say Debian Stretch works great, no flags needed, compiled with stock GCC.  Thanks for updating the miner!

**Comment by nioroso-x3** [comment]
@ethancedrik, please compile gcc 6.3.0, and compile the miner with it without setting any cmake flags.

**Comment by night2417** [comment]
please help me again
[root@localhost local]# cd xmr-stak/
[root@localhost xmr-stak]# mkdir build
[root@localhost xmr-stak]# cd build/
[root@localhost build]# ls
[root@localhost build]# cmake .. -DOpenSSL_ENABLE=OFF -DMICROHTTPD_ENABLE=OFF -DCMAKE_LINK_STATIC=ON
-- The C compiler identification is GNU 6.4.1
-- The CXX compiler identification is GNU 6.4.1
-- Check for working C compiler: /opt/at10.0/bin/cc
-- Check for working C compiler: /opt/at10.0/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/at10.0/bin/c++
-- Check for working CXX compiler: /opt/at10.0/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
**CMake Error at CMakeLists.txt:359 (message):
  hwloc NOT found: use `-DHWLOC_ENABLE=OFF` to build without hwloc support**


-- Configuring incomplete, errors occurred!
See also "/usr/local/xmr-stak/build/CMakeFiles/CMakeOutput.log".
See also "/usr/local/xmr-stak/build/CMakeFiles/CMakeError.log".
[root@localhost build]#

-DHWLOC_ENABLE=OFF Adding an option causes an error in make install.

**Comment by ethancedrik** [comment]
It looks like you’re missing the hwloc-devel package

**Comment by gcy000** [comment]
Manage to get it working for ubuntu 16.04 LTS on power8 
Summary steps in link: 
https://openpowerblog.wordpress.com/tag/ppc64le/

1. Running on power S822LC 20-core : smt=4 - hashrate 3790
Eg: on even number cpu 0,2,4,-158
{ "low_power_mode" : false, "be_mode" : false, "affine_to_cpu" : 0 },
{ "low_power_mode" : false, "be_mode" : false, "affine_to_cpu" : 2 },
.....
 { "low_power_mode" : false, "be_mode" : false, "affine_to_cpu" : 158},
 
2. Running on power S822 VIOS 16-core with default config on smt=4 - hashrate 3090

**Comment by tayore** [comment]
I can confirm that I compiled the new fork  on power8 ubuntu 16 LTS  but something is messed up.
strace output is the following;  I use at10



open("/usr/lib/powerpc64le-linux-gnu/libgpg-error.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/lib/powerpc64le-linux-gnu/libgpg-error.so.0", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0\25\0\1\0\0\0\340%\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=133512, ...}) = 0
mmap(NULL, 197168, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x3fff87ac0000
mmap(0x3fff87ae0000, 131072, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x10000) = 0x3fff87ae0000
close(3)                                = 0
open("/usr/lib/powerpc64le-linux-gnu/libffi.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0\25\0\1\0\0\0\300\24\0\0\0\0\0\0"..., 832) = 832
fstat(3, {st_mode=S_IFREG|0644, st_size=67808, ...}) = 0
mmap(NULL, 132504, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x3fff87a90000
mmap(0x3fff87aa0000, 131072, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0) = 0x3fff87aa0000
close(3)                                = 0
writev(2, [{"./xmr-stak", 10}, {": ", 2}, {"/opt/at10.0/lib64/power8/libcryp"..., 43}, {": ", 2}, {"no version information available"..., 57}, {"\n", 1}], 6./xmr-stak: /opt/at10.0/lib64/power8/libcrypto.so.1.0.0: no version information available (required by ./xmr-stak)
) = 115
writev(2, [{"./xmr-stak", 10}, {": ", 2}, {"/usr/lib/powerpc64le-linux-gnu/l"..., 45}, {": ", 2}, {"version `GLIBCXX_3.4.22' not fou"..., 59}, {"\n", 1}], 6./xmr-stak: /usr/lib/powerpc64le-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.22' not found (required by ./xmr-stak)
) = 119
writev(2, [{"./xmr-stak", 10}, {": ", 2}, {"/opt/at10.0/lib64/power8/libssl."..., 40}, {": ", 2}, {"no version information available"..., 57}, {"\n", 1}], 6./xmr-stak: /opt/at10.0/lib64/power8/libssl.so.1.0.0: no version information available (required by ./xmr-stak)
) = 112
exit_group(1)                           = ?
+++ exited with 1 +++

**Comment by tayore** [comment]
Compiled on  gcc version 7.3.1 20180207 (Advance-Toolchain-at11.0) [ibm/gcc-7-branch revision 257393] (GCC)  on ubuntu 16 LTS....

It seems  to be working with at11 instead at10, but  we need to focus on smt and cpu.txt
